src/core/mavsdk_impl.cpp

```diff
@@ -110,8 +110,71 @@ std::vector<std::shared_ptr<System>> MavsdkImpl::systems() const
     return systems_result;
 }
 
-void MavsdkImpl::receive_message(mavlink_message_t& message)
+void MavsdkImpl::forward_message(mavlink_message_t& message, Connection* connection)
+{
+    /**
+     * @brief forward_message Function implementing Mavlink routing rules.
+     * See https://mavlink.io/en/guide/routing.html
+     *
+     * This function was adapted from PX4 Firmware v1.11.3
+     * https://github.com/PX4/PX4-Autopilot/blob/v1.11.3/src/modules/mavlink/mavlink_main.cpp
+     */
+    const mavlink_msg_entry_t* meta = mavlink_get_msg_entry(message.msgid);
+
+    bool forward_heartbeats_enabled = false;
+    int target_system_id = 0;
+    int target_component_id = 0;
+
+    // might be nullptr if message is unknown
+    if (meta) {
+        // Extract target system and target component if set
+        if (meta->flags & MAV_MSG_ENTRY_FLAG_HAVE_TARGET_SYSTEM) {
+            target_system_id = (_MAV_PAYLOAD(&message))[meta->target_system_ofs];
+        }
+
+        if (meta->flags & MAV_MSG_ENTRY_FLAG_HAVE_TARGET_COMPONENT) {
+            target_component_id = (_MAV_PAYLOAD(&message))[meta->target_component_ofs];
+        }
+    }
+
+    // If it's a message only for us, we keep it, otherwise, we forward it.
+    const bool targeted_only_at_us =
+        (target_system_id == get_own_system_id() && target_component_id == get_own_component_id());
+
+    // We don't forward heartbeats unless it's specifically enabled.
+    const bool heartbeat_check_ok =
+        (message.msgid != MAVLINK_MSG_ID_HEARTBEAT || forward_heartbeats_enabled);
+
+    if (!targeted_only_at_us && heartbeat_check_ok) {
+        std::lock_guard<std::mutex> lock(_connections_mutex);
+
+        uint8_t successful_emissions = 0;
+        for (auto it = _connections.begin(); it != _connections.end(); ++it) {
+            if ((*it).get() == connection || !(**it).do_forward_messages()) {
+                continue;
+            }
+            if ((**it).send_message(message)) {
+                successful_emissions++;
+            }
+        }
+        if (successful_emissions == 0) {
+            LogErr() << "forward fail";
+        }
+    }
+}
+
+void MavsdkImpl::receive_message(mavlink_message_t& message, Connection* connection)
 {
+    /** @note: Forward message if option is enabled and multiple interfaces are connected.
+     *  Performs message forwarding checks for every messages if message forwarding
+     *  is enabled on at least one connection, and in case of a single forwarding connection,
+     *  we check that it is not the one which received the current message.
+     */
+    if (_connections.size() > 1 && connection->forwarding_connections_count() > 0 &&
+        (connection->forwarding_connections_count() && connection->do_forward_messages())) {
```

Comment:
You check twice if `connection->forwarding_connections_count()` but once should be enough, and you don't need the parentheses in the second part since it's all `&&`.

So what about?
```suggestion
    if (_connections.size() > 1 && connection->forwarding_connections_count() > 0 &&
        connection->do_forward_messages()) {
```

Reply:
Well I think I made a mistake, the conditions should be:
1. At least 2 connections.
2. At least 1 forwarding connection.
3. [If only one forwarding connection, it should not be the current one] = [At least 2 forwarding connections or current connection is not forwarding].

Reply:
I don't understand 3. :confused:

Reply:
Well if we have 3 connections: C1, C2 (forwarding) and C3.
If we receive a message from C2, we do not need to enter `forward_message` and performs all the checks as the all other connections do not forward messages.

Thus the new conditions are:
```
if (_connections.size() > 1 && 
    connection->forwarding_connections_count() > 0 &&
    (connection->forwarding_connections_count() > 1 || !connection->should_forward_messages())) 
{
    forward_message(message, connection);
}
```

Thus the 3rd condition checks that:
- Either we have at least 2 forwarding connections.
- Or (if there is only one forwarding connection) the connection from which we receive the message is not the the only forwarding connection.

Do you see what I mean ?

Reply:
Aha, so the rules for your example would be as follows:
C1 not forwarding
C2 forwarding
C3 not forwarding

Message arriving at C1 -> forwarded to C2
Message arriving at C2 -> not forwarded
Message arriving at C3 -> forwarded to C2

Is that correct?

Reply:
Yep, that was the idea !

Reply:
Ok, I think that's fine but we'll have to document that in https://github.com/mavlink/MAVSDK-docs/.